### ISSUE TITLE
AWS: Cannot kube-up in us-east-1 region.

### ISSUE THREAD

**Issue by iterion** [kubernetes/kubernetes#7962]
I was successfully able to get a cluster to start up in us-west-2, the default region for kube-up, but for some reason I'm not successfully able to start up a cluster in us-east-1.

I'm not seeing anything in the logs that seems to be a huge problem, so I'm not really sure how to resolve this.

Kubelet log here:
```
server.go:236] Using root directory: /var/lib/kubelet
manager.go:113] cAdvisor running in container: "/"
fs.go:87] Filesystem partitions: map[/dev/disk/by-uuid/5ecf1d67-1075-4dcc-bd94-0d4e17df997b:{mountpoint:/ major:202 minor:1}]
manager.go:134] Machine: {NumCores:2 CpuFrequency:2900098 MemoryCapacity:3946876928 MachineID:19b6e48dcced68b7e31ea64054cf9625 SystemUUID:EC245D6C-4926-B64C-790D-351D952F7CEA BootID:5dd71066-9db0-4200-8332-bbed54bf0e14 Filesystems:[{Device:/dev/disk/by-uuid/5ecf1d67-1075-4dcc-bd94-0d4e17df997b Capacity:8309932032}] DiskMap:map[252:0:{Name:dm-0 Major:252 Minor:0 Size:107374182400 Scheduler:none} 252:1:{Name:dm-1 Major:252 Minor:1 Size:10737418240 Scheduler:none} 202:0:{Name:xvda Major:202 Minor:0 Size:8589934592 Scheduler:deadline}] NetworkDevices:[{Name:cbr0 MacAddress:d2:72:92:85:25:b8 Speed:0 Mtu:1460} {Name:eth0 MacAddress:0e:34:0b:ac:ce:17 Speed:10000 Mtu:9001}] Topology:[{Id:0 Memory:3946876928 Cores:[{Id:0 Threads:[0 1] Caches:[{Size:32768 Type:Data Level:1} {Size:32768 Type:Instruction Level:1} {Size:262144 Type:Unified Level:2}]}] Caches:[{Size:26214400 Type:Unified Level:3}]}]}
manager.go:141] Version: {KernelVersion:3.16.0-30-generic ContainerOsVersion:Ubuntu 14.10 DockerVersion:1.6.0 CadvisorVersion:0.13.0}
factory.go:228] Registering Docker factory
factory.go:64] Registering Raw factory
manager.go:897] Started watching for new ooms in manager
oomparser.go:199] OOM parser using kernel log file: "/var/log/kern.log"
manager.go:219] Starting recovery of all containers
manager.go:224] Recovery completed
plugins.go:69] No cloud provider specified.
server.go:251] Successfully initialized cloud provider: "" from the config file: ""
server.go:264] Using self-signed cert (/var/run/kubernetes/kubelet.crt, /var/run/kubernetes/kubelet.key)
docker.go:258] Connecting to docker on unix:///var/run/docker.sock
server.go:452] No api server defined - no events will be sent to API server.
plugins.go:56] Registering credential provider: .dockercfg
server.go:477] Started kubelet
kubelet.go:600] No api server defined - no node status update will be sent.
server.go:63] Starting to listen on 0.0.0.0:10250
event.go:203] Event(api.ObjectReference{Kind:"Node", Namespace:"", Name:"ip-172-20-0-9", UID:"ip-172-20-0-9", APIVersion:"", ResourceVersion:"", FieldPath:""}): reason: 'starting' Starting kubelet.
kubelet.go:609] Running in container "/kubelet"
status_manager.go:56] Starting to sync pod status with apiserver
kubelet.go:1326] Starting kubelet main sync loop.
kubelet.go:1281] error getting node: node ip-172-20-0-9 not found
kubelet.go:1281] error getting node: node ip-172-20-0-9 not found
kubelet.go:1281] error getting node: node ip-172-20-0-9 not found
kubelet.go:1281] error getting node: node ip-172-20-0-9 not found
kubelet.go:1281] error getting node: node ip-172-20-0-9 not found
kubelet.go:1281] error getting node: node ip-172-20-0-9 not found
```

I'm not really sure what else to look at, I'm happy to do more investigation, but not really sure where to look. Most others log files are empty with the exception of kube-addons.log which is failing to start addons due to the unstarted api server.

**Comment by dchen1107** [comment]
cc/ @justinsb Could you please help us with this?

**Comment by justinsb** [comment]
Sure ... this sounds like a name problem - as I recall the name is supposed to be ip-172-20-0-9.amazonaws.com (or similar), not just ip-172-20-0-9.

I'll try in us-east-1.  What revision of the software is this @iterion ?

**Comment by iterion** [comment]
I built from the latest master, but I have seen the same issue on recent releases.

**Comment by justinsb** [comment]
Testing now...  I wonder if this is related to #7910 and #7967

**Comment by timothysc** [comment]
@justinsb @iterion if you entries in etcd were the fqdn of the machines then you would have been bit by the same issue as us.

**Comment by iterion** [comment]
@justinsb I was skeptical about https://github.com/GoogleCloudPlatform/kubernetes/blob/master/cluster/aws/templates/salt-minion.sh#L37

Running that curl returns "ip-172-20-0-9" and not "ip-172-20-0-9.ec2.internal" as we might expect?

**Comment by iterion** [comment]
Nvm, I tried just appending ".ec2.internal" to no avail.

**Comment by justinsb** [comment]
I just tried it, and I was able to bring up a cluster in us-east with the latest code.

I checked on my kubelet on the master, and did see your error:
```
 Cannot get host IP: cannot get node: node ip-172-20-0-9 not found
```

So I think that error is a red herring.  Of course we should fix it, though I don't know if the problem is that it is missing the .ec2.internal suffix or whether the problem is that we're not registering the master as a node at all.

Do you see the API server in `docker ps` (or `sudo docker ps`) on the master.  I suspect it isn't running.  Or maybe it isn't configured on your kubelet - what does `ps -ef | grep kubelet` show?

**Comment by iterion** [comment]
Nothing seems to be running:

```
ubuntu@ip-172-20-0-9:~$ sudo docker ps
sudo: unable to resolve host ip-172-20-0-9
CONTAINER ID        IMAGE               COMMAND             CREATED             STATUS              PORTS               NAMES
```

```
ubuntu@ip-172-20-0-9:~$ ps -ef | grep kubelet
root      7428     1  0 00:21 ?        00:00:06 /usr/local/bin/kubelet
ubuntu   31045 31000  0 00:46 pts/4    00:00:00 grep --color=auto kubelet
```

**Comment by justinsb** [comment]
That is odd... here is my output:

```
ubuntu@ip-172-20-0-9:~$ sudo docker ps
CONTAINER ID        IMAGE                                                                               COMMAND                CREATED             STATUS              PORTS               NAMES
b4d89b0d2313        gcr.io/google_containers/kube-controller-manager:6c0743f48f87912e528696e490e5e3f4   "/bin/sh -c '/usr/lo   3 hours ago         Up 3 hours                              k8s_kube-controller-manager.d138c2fa_kube-controller-manager-ip-172-20-0-9_default_a0837735dd9f38536aa369cd556bef09_857cfa15
...
```

Note no `sudo: unable to resolve host ip-172-20-0-9`

```
ps -ef | grep kubelet
root      6167     1  1 May08 ?        00:02:20 /usr/local/bin/kubelet --api_servers=https://172.20.0.9 --config=/etc/kubernetes/manifests --allow_privileged=False --v=2 --cluster_dns=10.0.0.10 --cluster_domain=kubernetes.local
ubuntu   23104 23039  0 01:27 pts/4    00:00:00 grep --color=auto kubelet
```

Note that I have arguments to kubelet (a lot of them!)  The lack of arguments matches with the " No api server defined - no node status update will be sent." message in your kubelet log (which I don't have)

On the "unable to resolve host ip-172-20-0-9" error, I can `ping ip-172-20-0-9`.  It isn't in hostname, but rather
```
cat /etc/resolv.conf
# Dynamic resolv.conf(5) file for glibc resolver(3) generated by resolvconf(8)
#     DO NOT EDIT THIS FILE BY HAND -- YOUR CHANGES WILL BE OVERWRITTEN
nameserver 172.20.0.2
search ec2.internal
```

I wonder if you can't `ping ip-172-20-0-9`; a possible reason would be that your VPC doesn't have name resolution enabled.  Did the script create it for you?

Are there any hints in `cat /var/log/salt/minion`?   Here is mine:

```
015-05-08 21:54:18,784 [salt.loaded.int.states.file][WARNING ] Neither 'source' nor 'contents' nor 'contents_pillar' nor 'contents_grains' was defined, yet 'replace' was set to 'True'. As there is no source to replace the file with, 'replace' has been set to 'False' to avoid reading the file unnecessarily
2015-05-08 21:54:19,145 [salt.loaded.int.states.file][WARNING ] Neither 'source' nor 'contents' nor 'contents_pillar' nor 'contents_grains' was defined, yet 'replace' was set to 'True'. As there is no source to replace the file with, 'replace' has been set to 'False' to avoid reading the file unnecessarily
2015-05-08 21:54:19,155 [salt.loaded.int.states.file][WARNING ] Neither 'source' nor 'contents' nor 'contents_pillar' nor 'contents_grains' was defined, yet 'replace' was set to 'True'. As there is no source to replace the file with, 'replace' has been set to 'False' to avoid reading the file unnecessarily
2015-05-08 21:54:26,634 [salt.loaded.int.module.cmdmod][ERROR   ] Command ['service', 'kube-master-addons', 'status'] failed with return code: 1
2015-05-08 21:54:26,635 [salt.loaded.int.module.cmdmod][ERROR   ] output:
2015-05-08 21:54:27,054 [salt.loaded.int.module.cmdmod][ERROR   ] Command ['service', 'kube-addons', 'status'] failed with return code: 1
2015-05-08 21:54:27,054 [salt.loaded.int.module.cmdmod][ERROR   ] output:
2015-05-08 21:54:29,622 [salt.loaded.int.module.cmdmod][ERROR   ] Command '/sbin/iptables -t nat -C POSTROUTING -o eth0 -j MASQUERADE \\! -d 10.0.0.0/8' failed with return code: 1
2015-05-08 21:54:29,622 [salt.loaded.int.module.cmdmod][ERROR   ] output: iptables: No chain/target/match by that name.
2015-05-08 21:54:29,797 [salt.loaded.int.module.cmdmod][ERROR   ] Command '/sbin/iptables -t nat -C POSTROUTING -o eth0 -j MASQUERADE \\! -d 10.0.0.0/8' failed with return code: 1
2015-05-08 21:54:29,797 [salt.loaded.int.module.cmdmod][ERROR   ] output: iptables: No chain/target/match by that name.
2015-05-08 21:54:29,812 [salt.loaded.int.module.cmdmod][ERROR   ] Command '/sbin/iptables -t nat -C POSTROUTING -o eth0 -j MASQUERADE \\! -d 10.0.0.0/8' failed with return code: 1
2015-05-08 21:54:29,812 [salt.loaded.int.module.cmdmod][ERROR   ] output: iptables: No chain/target/match by that name.
2015-05-08 21:54:29,827 [salt.loaded.int.module.cmdmod][ERROR   ] Command '/sbin/iptables -t nat -C POSTROUTING -o eth0 -j MASQUERADE \\! -d 10.0.0.0/8' failed with return code: 1
2015-05-08 21:54:29,827 [salt.loaded.int.module.cmdmod][ERROR   ] output: iptables: No chain/target/match by that name.
2015-05-08 21:54:29,842 [salt.loaded.int.module.cmdmod][ERROR   ] Command '/sbin/iptables -t nat -C POSTROUTING -o eth0 -j MASQUERADE \\! -d 10.0.0.0/8' failed with return code: 1
2015-05-08 21:54:29,843 [salt.loaded.int.module.cmdmod][ERROR   ] output: iptables: No chain/target/match by that name.
2015-05-08 21:54:42,532 [salt.loaded.int.module.cmdmod][ERROR   ] Command ['service', 'kubelet', 'status'] failed with return code: 3
2015-05-08 21:54:42,532 [salt.loaded.int.module.cmdmod][ERROR   ] output:  * kubelet is not running
2015-05-08 21:59:01,412 [salt.loaded.int.states.file][WARNING ] Neither 'source' nor 'contents' nor 'contents_pillar' nor 'contents_grains' was defined, yet 'replace' was set to 'True'. As there is no source to replace the file with, 'replace' has been set to 'False' to avoid reading the file unnecessarily
2015-05-08 21:59:03,019 [salt.loaded.int.states.file][WARNING ] Neither 'source' nor 'contents' nor 'contents_pillar' nor 'contents_grains' was defined, yet 'replace' was set to 'True'. As there is no source to replace the file with, 'replace' has been set to 'False' to avoid reading the file unnecessarily
2015-05-08 21:59:03,029 [salt.loaded.int.states.file][WARNING ] Neither 'source' nor 'contents' nor 'contents_pillar' nor 'contents_grains' was defined, yet 'replace' was set to 'True'. As there is no source to replace the file with, 'replace' has been set to 'False' to avoid reading the file unnecessarily
2015-05-08 21:59:03,271 [salt.loaded.int.module.cmdmod][ERROR   ] Command ['service', 'kube-master-addons', 'status'] failed with return code: 1
2015-05-08 21:59:03,272 [salt.loaded.int.module.cmdmod][ERROR   ] output:
2015-05-08 21:59:03,419 [salt.loaded.int.module.cmdmod][ERROR   ] Command ['service', 'kube-addons', 'status'] failed with return code: 1
2015-05-08 21:59:03,420 [salt.loaded.int.module.cmdmod][ERROR   ] output:
```

**Comment by iterion** [comment]
VPC was created by the script and dns support and dns hostnames are both turned on.

Interestingly, I'm unable to ping ip-172-20-0-9 and my resolv.conf is missing the search line:

```
# Dynamic resolv.conf(5) file for glibc resolver(3) generated by resolvconf(8)
#     DO NOT EDIT THIS FILE BY HAND -- YOUR CHANGES WILL BE OVERWRITTEN
nameserver 172.20.0.2
```

`cat /var/log/salt/minion` looks pretty close, but noticed some extra lines re: Jinja:
```
2015-05-09 00:21:33,825 [salt.loaded.int.states.file][WARNING ] Neither 'source' nor 'contents' nor 'contents_pillar' nor 'contents_grains' was defined, yet 'replace' was set to 'True'. As there is no source to replace the file with, 'replace' has been set to 'False' to avoid reading the file unnecessarily
2015-05-09 00:21:34,107 [salt.loaded.int.states.file][WARNING ] Neither 'source' nor 'contents' nor 'contents_pillar' nor 'contents_grains' was defined, yet 'replace' was set to 'True'. As there is no source to replace the file with, 'replace' has been set to 'False' to avoid reading the file unnecessarily
2015-05-09 00:21:34,114 [salt.loaded.int.states.file][WARNING ] Neither 'source' nor 'contents' nor 'contents_pillar' nor 'contents_grains' was defined, yet 'replace' was set to 'True'. As there is no source to replace the file with, 'replace' has been set to 'False' to avoid reading the file unnecessarily
2015-05-09 00:21:40,035 [salt.loaded.int.module.cmdmod][ERROR   ] Command ['service', 'kube-master-addons', 'status'] failed with return code: 1
2015-05-09 00:21:40,035 [salt.loaded.int.module.cmdmod][ERROR   ] output:
2015-05-09 00:21:40,512 [salt.loaded.int.module.cmdmod][ERROR   ] Command ['service', 'kube-addons', 'status'] failed with return code: 1
2015-05-09 00:21:40,512 [salt.loaded.int.module.cmdmod][ERROR   ] output:
2015-05-09 00:21:42,437 [salt.loaded.int.module.cmdmod][ERROR   ] Command '/sbin/iptables -t nat -C POSTROUTING -o eth0 -j MASQUERADE \\! -d 10.0.0.0/8' failed with return code: 1
2015-05-09 00:21:42,437 [salt.loaded.int.module.cmdmod][ERROR   ] output: iptables: No chain/target/match by that name.
2015-05-09 00:21:42,603 [salt.loaded.int.module.cmdmod][ERROR   ] Command '/sbin/iptables -t nat -C POSTROUTING -o eth0 -j MASQUERADE \\! -d 10.0.0.0/8' failed with return code: 1
2015-05-09 00:21:42,604 [salt.loaded.int.module.cmdmod][ERROR   ] output: iptables: No chain/target/match by that name.
2015-05-09 00:21:42,618 [salt.loaded.int.module.cmdmod][ERROR   ] Command '/sbin/iptables -t nat -C POSTROUTING -o eth0 -j MASQUERADE \\! -d 10.0.0.0/8' failed with return code: 1
2015-05-09 00:21:42,618 [salt.loaded.int.module.cmdmod][ERROR   ] output: iptables: No chain/target/match by that name.
2015-05-09 00:21:42,632 [salt.loaded.int.module.cmdmod][ERROR   ] Command '/sbin/iptables -t nat -C POSTROUTING -o eth0 -j MASQUERADE \\! -d 10.0.0.0/8' failed with return code: 1
2015-05-09 00:21:42,633 [salt.loaded.int.module.cmdmod][ERROR   ] output: iptables: No chain/target/match by that name.
2015-05-09 00:21:42,647 [salt.loaded.int.module.cmdmod][ERROR   ] Command '/sbin/iptables -t nat -C POSTROUTING -o eth0 -j MASQUERADE \\! -d 10.0.0.0/8' failed with return code: 1
2015-05-09 00:21:42,647 [salt.loaded.int.module.cmdmod][ERROR   ] output: iptables: No chain/target/match by that name.
2015-05-09 00:21:59,334 [salt.state       ][ERROR   ] Unable to manage file: Jinja variable list object has no element 0
2015-05-09 00:22:00,546 [salt.loaded.int.module.cmdmod][ERROR   ] Command ['service', 'kubelet', 'status'] failed with return code: 3
2015-05-09 00:22:00,547 [salt.loaded.int.module.cmdmod][ERROR   ] output:  * kubelet is not running
2015-05-09 00:26:30,345 [salt.loaded.int.states.file][WARNING ] Neither 'source' nor 'contents' nor 'contents_pillar' nor 'contents_grains' was defined, yet 'replace' was set to 'True'. As there is no source to replace the file with, 'replace' has been set to 'False' to avoid reading the file unnecessarily
2015-05-09 00:26:30,623 [salt.loaded.int.states.file][WARNING ] Neither 'source' nor 'contents' nor 'contents_pillar' nor 'contents_grains' was defined, yet 'replace' was set to 'True'. As there is no source to replace the file with, 'replace' has been set to 'False' to avoid reading the file unnecessarily
2015-05-09 00:26:30,630 [salt.loaded.int.states.file][WARNING ] Neither 'source' nor 'contents' nor 'contents_pillar' nor 'contents_grains' was defined, yet 'replace' was set to 'True'. As there is no source to replace the file with, 'replace' has been set to 'False' to avoid reading the file unnecessarily
2015-05-09 00:26:30,746 [salt.loaded.int.module.cmdmod][ERROR   ] Command ['service', 'kube-master-addons', 'status'] failed with return code: 1
2015-05-09 00:26:30,746 [salt.loaded.int.module.cmdmod][ERROR   ] output:
2015-05-09 00:26:30,848 [salt.loaded.int.module.cmdmod][ERROR   ] Command ['service', 'kube-addons', 'status'] failed with return code: 1
2015-05-09 00:26:30,849 [salt.loaded.int.module.cmdmod][ERROR   ] output:
2015-05-09 00:26:31,150 [salt.state       ][ERROR   ] Unable to manage file: Jinja variable list object has no element 0
```

**Comment by iterion** [comment]
I think it might be the DHCP Options Set. I have another VPC in east that already had a DHCP Option Set and it appears that the kubernetes cluster is using that same set.

When I looked in west I found that the set was configured as such:
```
domain-name = us-west-2.compute.internal
domain-name-servers = AmazonProvidedDNS
```

Whereas in east, my DHCP option set just looked like this:
```
domain-name-servers = AmazonProvidedDNS
```

Perhaps we should be explicitly creating a new DHCP option set to avoid conflicts with existing VPCs?

I'd be happy to create a PR if this turns out to be the issue and if you think the above is a good solution.

**Comment by justinsb** [comment]
@iterion That is almost certainly it - nice find!  I don't fully understand how this actually happened (I guess a new VPC defaults to whatever DHCP option set already exists, although that isn't what the docs say), but I would welcome a patch that created a working DHCP option set when creating a new VPC, and/or warned if a VPC had a DHCP option set which didn't include a domain-name.

**Comment by iterion** [comment]
Yep, cluster in us-east-1 is up and running! Thanks for all of the troubleshooting help!

Another option might be to set the DHCP option set to default, which should be configured correctly. It doesn't seem like you can modify the default, so it might be a safe choice. Plus, we don't have to set up the option set, we can just assign it through the [aws cli](http://docs.aws.amazon.com/cli/latest/reference/ec2/associate-dhcp-options.html). I'll do some experimentation and make a PR with what seems to be the most sane.

The documentation for option sets seems contradictory, in one place it says it will create a new option set and in another it says it will use the default. It seems it does a little bit of both. :)

**Comment by alex-mohr** [comment]
#8255 should resolve once it merges.

**Comment by francoran** [comment]
Unfortunately this issue is still exist for me.
Even tough a DHCP option is created for the Kubernetes VPC I've get an endless loop.

>ubuntu@ip-172-20-0-9:~$ sudo docker ps
>sudo: unable to resolve host ip-172-20-0-9
>FATA[0002] Get http:///var/run/docker.sock/v1.18/containers/json: read unix /var/run/docker.sock

**Comment by francoran** [comment]
@alex-mohr 
I figured it out,
The issue was that as part of the dhcp set creation it didn't set a domain-name to be "ec2.interna" as described here - http://docs.aws.amazon.com/AmazonVPC/latest/UserGuide/VPC_DHCP_Options.html
and only set the domain-server to AWS defaults.
Once I changed that dhcp set it worked perfectly.

I can see you added a function for that issue in #8255, it might not merged to the master?

Thanks
